### ISSUE TITLE
SiteImprove reports WCAG 2.0 violation: Element ID is not unique

### ISSUE THREAD

**Issue by tatwell** [chapmanu/cascade-assets#310]
Trello Cards: 

- https://trello.com/c/YxTEzOLv
- https://trello.com/c/tNeLAuXj

This issue started to be reported for almost every WWW page after we rolled out the new OmniNav. Here's the issue in SiteImprove (account access required):

- https://my2.siteimprove.com/Accessibility/10993/Pages/Issue?CheckId=33&CriterionChapter=4.1.1

Unfortunately, SiteImprove doesn't report which ID exactly is not unique. But we suspect it is associated with the OmniNav search form, which gets duplicated to accommodate desktop/mobile reflows.

**Comment by tatwell** [comment]
I think I've identified the duplicate element. It's not quite where I expected.

Here is the markup for the desktop version of OmniNav search form (as seen on [Chapman home page](https://www.chapman.edu/)):

```
<div class="cu-search-box">
  <div id="___gcse_0" dir="ltr">
    <form class="gsc-search-box" accept-charset="utf-8">
      <table cellspacing="0" cellpadding="0" class="gsc-search-box">
        <tbody>
          <tr>
            <td class="gsc-input">
              <input autocomplete="off" type="text" size="10" class=" gsc-input " name="search" title="search" id="gsc-i-id1" dir="ltr" spellcheck="false" placeholder="Search" style="outline: none; background: url(&quot;https://www.google.com/cse/static/images/1x/googlelogo_lightgrey_46x16dp.png&quot;) left center no-repeat rgb(255, 255, 255); text-indent: 48px;">
              <input type="hidden" name="bgresponse" id="bgresponse">
            </td>
            <td class="gsc-search-button">
              <input type="button" value="Search" class="gsc-search-button" title="search">
            </td>
            <td class="gsc-clear-button">
              <div class="gsc-clear-button" title="clear results">&nbsp;</div>
            </td>
          </tr>
        </tbody>
      </table>
      <table cellspacing="0" cellpadding="0" class="gsc-branding">
        <tbody>
          <tr>
            <td class="gsc-branding-user-defined"></td>
            <td class="gsc-branding-text">
              <div class="gsc-branding-text">powered by</div>
            </td>
            <td class="gsc-branding-img"><img src="https://www.google.com/cse/static/images/1x/googlelogo_grey_46x15dp.png" class="gsc-branding-img" srcset="https://www.google.com/cse/static/images/2x/googlelogo_grey_46x15dp.png 2x"></td>
          </tr>
        </tbody>
      </table>
    </form>
  </div>
</div>
```

Here's the mobile version of the form:

```
<div class="cu-search-box">
  <div id="___gcse_2" dir="ltr">
    <form class="gsc-search-box" accept-charset="utf-8">
      <table cellspacing="0" cellpadding="0" class="gsc-search-box">
        <tbody>
          <tr>
            <td class="gsc-input">
              <input autocomplete="off" type="text" size="10" class=" gsc-input " name="search" title="search" id="gsc-i-id2" dir="ltr" spellcheck="false" placeholder="Search" style="outline: none; background: url(&quot;https://www.google.com/cse/static/images/1x/googlelogo_lightgrey_46x16dp.png&quot;) left center no-repeat rgb(255, 255, 255); text-indent: 48px;">
              <input type="hidden" name="bgresponse" id="bgresponse">
            </td>
            <td class="gsc-search-button">
              <input type="button" value="Search" class="gsc-search-button" title="search">
            </td>
            <td class="gsc-clear-button">
              <div class="gsc-clear-button" title="clear results">&nbsp;</div>
            </td>
          </tr>
        </tbody>
      </table>
      <table cellspacing="0" cellpadding="0" class="gsc-branding">
        <tbody>
          <tr>
            <td class="gsc-branding-user-defined"></td>
            <td class="gsc-branding-text">
              <div class="gsc-branding-text">powered by</div>
            </td>
            <td class="gsc-branding-img"><img src="https://www.google.com/cse/static/images/1x/googlelogo_grey_46x15dp.png" class="gsc-branding-img" srcset="https://www.google.com/cse/static/images/2x/googlelogo_grey_46x15dp.png 2x"></td>
          </tr>
        </tbody>
      </table>
    </form>
  </div>
</div>
```

I was expecting the duplicate ID to be the parent element (e.g. `div#___gcse_0`). But it looks like it's the hidden input element with id `bgresponse` that's being duplicated.

**Comment by tatwell** [comment]
The problem with the duplicate `input#bgresponse` element is that that is an element that is automatically generated by Google's API and it doesn't offer a [configurable attribute](https://developers.google.com/custom-search/docs/element#supported_attributes) for changing the ID AFAICT.

If true, this means that to avoid the SI warning, we would need figure out a way to limit the search form to one instance and swap it's location according to the view context (mobile or not).

I want to review this with the product owner. I noticed in the SI dashboard that it is possible to mark issues as `wontfix`. The WAVE tool isn't reporting this issue. So if this story is simply to address this SI report, perhaps it will be satisfactory to handle this way.

**Comment by kaylaziegler** [comment]
If you login to the SiteImprove dashboard and go to the Accessibility section overview, one of the main issues under the Webmaster tab is this `Element ID is not unique` violation. If you click on that, it takes you to a page that lists all the pages where the error is occurring and you can see an individual report for each page that shows exactly where in the HTML it sees the duplicate IDs.

For example, this is the [report](http://accessibility.page.report/E133425B7B2E3ECBD2ED2746903F43B3/10993/23242233415#automatic/4.1.1/33) for the homepage:
![screen shot 2018-01-05 at 9 18 54 am](https://user-images.githubusercontent.com/11510966/34620472-6beaa56a-f1fa-11e7-86fd-ae775c6be97a.png)

This is the additional detail it gives for the occurrence we're most interested in here:
![screen shot 2018-01-05 at 9 26 12 am](https://user-images.githubusercontent.com/11510966/34620513-92225c1e-f1fa-11e7-95bd-ee3e659c55da.png)

In the DOM for the live homepage, you can see that the `bgresponse` elements with duplicate IDs are located within the google custom search code in both the utility nav and the primary nav. As @tatwell commented, the code with the duplicated IDs comes automatically from Google's custom search API and isn't something we put in there ourselves.

**Comment by tatwell** [comment]
Thanks @kaylaziegler. That's extremely helpful. And now that I double-check, I see I completely overlooked the redundant `<g id="icomoon-ignore"></g>` tags. They can be found in the OmniNav format here:

- https://github.com/chapmanu/cascade-assets/blob/3e04f7aa7c93c3fa52d410b3850ab59a88f44abe/.cascade-code/Chapman.edu/_components/omni_nav_v2/omni_nav_v2.vtl#L772

@mandythomas or @waidelacruz Do you have any idea why that `id` is included with `g` tag? I am going to try to remove it.

**Comment by tatwell** [comment]
The original story associated with this story has been completed:

- https://trello.com/c/tNeLAuXj/#comment-5a5e3872fbb2e48b4ced64c2

However, not all the SiteImprove issues have been eliminated. So a new story has been created to address those:

- https://trello.com/c/YxTEzOLv/725-omninav-causes-errors-in-siteimprove

I'm leaving the issue open but expect it to be closed with this new story.

**Comment by kaylaziegler** [comment]
## Update

I'm currently working on this story to address all accessibility issues related to the OmniNav: https://trello.com/c/YxTEzOLv

In regards to the duplicate ID error in SiteImprove that caused us to open this GitHub issue in the first place, it seems Google changed the HTML a little bit for the custom search function. So now we do not have the `bgresponse` duplicate ID problem that we had in January. To compare with [Tom's comment here](https://github.com/chapmanu/cascade-assets/issues/310#issuecomment-355380854), this is what the HTML looks like now:

```
<div class="cu-search-box">
    <div id="___gcse_2" dir="ltr">
        <form class="gsc-search-box" accept-charset="utf-8">
            <table cellspacing="0" cellpadding="0" class="gsc-search-box">
                <tbody>
                    <tr>
                        <td class="gsc-input">
                            <input autocomplete="off" type="text" size="10" class=" gsc-input " name="search" title="search" id="gsc-i-id2" dir="ltr" spellcheck="false" placeholder="Search" style="outline: none; background: url(&quot;https://www.google.com/cse/static/images/1x/googlelogo_lightgrey_46x16dp.png&quot;) left center no-repeat rgb(255, 255, 255); text-indent: 48px;">
                        </td>
                        <td class="gsc-search-button">
                            <input type="button" value="Search" class="gsc-search-button" title="search">
                        </td>
                        <td class="gsc-clear-button">
                            <div class="gsc-clear-button" title="clear results">&nbsp;</div>
                        </td>
                    </tr>
                </tbody>
            </table>
            <table cellspacing="0" cellpadding="0" class="gsc-branding">
                <tbody>
                    <tr>
                        <td class="gsc-branding-user-defined"></td>
                        <td class="gsc-branding-text">
                            <div class="gsc-branding-text">powered by</div>
                        </td>
                        <td class="gsc-branding-img"><img src="https://www.google.com/cse/static/images/1x/googlelogo_grey_46x15dp.png" class="gsc-branding-img" srcset="https://www.google.com/cse/static/images/2x/googlelogo_grey_46x15dp.png 2x"></td>
                    </tr>
                </tbody>
            </table>
        </form>
    </div>
</div>
```

You can see there is no longer a hidden `#bgresponse` element, which is what caused the duplicate ID error from SiteImprove.

However, I will still be modifying the JS so that the google custom search only gets loaded once in one place on the page, instead of twice. This will protect us in the future if the `bgresponse` element comes back for some reason, and reduces the number of accessibility issues on the pages that are caused by Google custom search code that we cannot modify.

**Comment by kaylaziegler** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [New OmniNav is causing accessibility errors on pages](https://trello.com/c/YxTEzOLv/725-new-omninav-is-causing-accessibility-errors-on-pages)
